compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -162,10 +162,18 @@ fn fmt_printer<'a, 'tcx>(infcx: &'a InferCtxt<'tcx>, ns: Namespace) -> FmtPrinte
         let mut infcx_inner = infcx.inner.borrow_mut();
         let ty_vars = infcx_inner.type_variables();
         let var_origin = ty_vars.var_origin(ty_vid);
-        if let TypeVariableOriginKind::TypeParameterDefinition(name, _) = var_origin.kind
+        if let TypeVariableOriginKind::TypeParameterDefinition(name, def_id) = var_origin.kind
             && !var_origin.span.from_expansion()
         {
-            Some(name)
+            let generics = infcx.tcx.generics_of(infcx.tcx.parent(def_id));
+            if let Some(idx) = generics.param_def_id_to_index(infcx.tcx, def_id)
+                && let generic_param_def = generics.param_at(idx as usize, infcx.tcx)
```

Comment:
```suggestion
            if let Some(idx) = generics.param_def_id_to_index(infcx.tcx, def_id)
                && let generic_param_def = generics.param_at(idx as usize, infcx.tcx)
```
please `unwrap` instead of using `if let`. Not getting `Some` here indicates a bug

Reply:
very astute! https://github.com/rust-lang/rust/issues/113610